### PR TITLE
fix: Dropdown option gets overlapped on the ModalComponent

### LINKED ISSUE #1000: [Bug] Dropdown : Dropdown option gets overlapped on the Modal
## Steps to reproduce

1. Place a dropdown. In the property pane, make it a multi-select dropdown.
2. In the property pane, configure `onOptionChange` to open a modal. 
3. Change an option in the multi-select dropdown

## Observation

Dropdown options get overlapped on the Modal

## Expected behaviour

Dropdown option must not get overlapped on the Modal. The multi-select dropdown must close itself.

## Video reproducing this bug:
[https://www.loom.com/share/eba7375d05c645508cd21b631db1d14d](https://www.loom.com/share/eba7375d05c645508cd21b631db1d14d)

### PR DESCRIPTION
**Pull request by Tooluloope**
By reducing the portal z-index and increasing that of the Modal solved the issue

Fixes #1000 

## Type of change

- Bug fix (non-breaking change which fixes an issue)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:red_circle: Total coverage has decreased</summary>

    // Code coverage diff between base branch:release and head branch: fix/dropdown-overlap-modal 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :red_circle: | total | 53.27 **(-0.02)** | 35.26 **(-0.01)** | 31.86 **(-0.02)** | 53.78 **(-0.02)**
 :red_circle: | app/client/src/components/designSystems/appsmith/Dropdown.tsx | 31.82 **(-1.51)** | 0 **(0)** | 0 **(0)** | 31.82 **(-1.51)**
 :red_circle: | app/client/src/components/designSystems/blueprint/DropdownComponent.tsx | 32.1 **(-0.79)** | 22.64 **(0.42)** | 3.23 **(-0.22)** | 32.05 **(-1.28)**
 :red_circle: | app/client/src/widgets/DropdownWidget.tsx | 42.37 **(-3.08)** | 0 **(0)** | 38.46 **(-3.21)** | 43.64 **(-3.42)**</details>

### CODE REVIEW

**Comment by marks0351** [comment]
@Tooluloope looks like when the modal is open, the user cannot get to the entity explorer to add widgets to the modal.
also, u cud try the usePortal option provided by blueprint to fix this problem.

**Comment by Tooluloope** [comment]
> @Tooluloope looks like when the modal is open, the user cannot get to the entity explorer to add widgets to the modal.
> also, u cud try the usePortal option provided by blueprint to fix this problem.

@marks0351 to get you straight, the modal isn't meant to cover the sidebar(Entity Explorer) when the modal is opened?
Screenshot 2021-05-09 at 13 48 36

**Comment by somangshu** [comment]
> > @Tooluloope looks like when the modal is open, the user cannot get to the entity explorer to add widgets to the modal.
> > also, u cud try the usePortal option provided by blueprint to fix this problem.
> 
> @marks0351 to get you straight, the modal isn't meant to cover the sidebar(Entity Explorer) when the modal is opened?
> Screenshot 2021-05-09 at 13 48 36

@Tooluloope that is right! As @marks0351 mentioned, the ability to drag new widgets inside the modal widget viewport is required. If the entity explorer is covered by the modal overlay, this won't be possible.

**Review by somangshu — CHANGES_REQUESTED** [review]
@Tooluloope this doesn't seem right, The overlay is now only covering half the canvas and also allows scroll. @marks0351 kindly look into this and help up out with the right direction.

Also @Tooluloope we need to link the issue which has additionally been merged here.

File: app/client/src/components/designSystems/blueprint/DropdownComponent.tsx [hunk 1]
```diff
@@ -230,6 +230,14 @@ const StyledMultiDropDown = styled(MultiDropDown)<{
 `;
 
 class DropDownComponent extends React.Component<DropDownComponentProps> {
+  private _menu = React.createRef<HTMLDivElement>();
+
+  isModalWidgetParent = () => {
+    if (Boolean(this._menu.current?.closest(".bp3-modal-widget"))) {
```
Comment (marks0351) [inline app/client/src/components/designSystems/blueprint/DropdownComponent.tsx:236]: Context about other widgets should not be present in the design system. u can add portalContainer as an input prop for this component and set it from the place where dropdown is used(DropdownWidget).

File: app/client/src/components/designSystems/blueprint/ModalComponent.tsx [hunk 2]
```diff
@@ -75,37 +75,52 @@ export function ModalComponent(props: ModalComponentProps) {
       modalContentRef.current?.scrollTo({ top: 0, behavior: "smooth" });
     }
   }, [props.scrollContents]);
+  console.log(props.height, props.width, props.left, props.top);
```
Comment (marks0351) [inline app/client/src/components/designSystems/blueprint/ModalComponent.tsx:78]: remove console.log

File: app/client/src/components/designSystems/blueprint/ModalComponent.tsx [hunk 3]
```diff
@@ -75,37 +75,52 @@ export function ModalComponent(props: ModalComponentProps) {
...
   return (
-    <Container
-      height={props.height}
-      left={props.left}
-      top={props.top}
-      width={props.width}
-      zIndex={props.zIndex !== undefined ? props.zIndex : Layers.modalWidget}
+    <Overlay
```
Comment (marks0351) [inline app/client/src/components/designSystems/blueprint/ModalComponent.tsx:80]: I'd suggest not to make changes to the modal. I think the modal portal can still stay at the body level. But the dropdown widget inside a modal should make use of the modal as the portal. let me know if that doesn't make any sense.

**Review by marks0351 — CHANGES_REQUESTED** [review]

File: app/client/src/components/designSystems/blueprint/ModalComponent.tsx (hunk 3 above)
Comment (Tooluloope) [reply app/client/src/components/designSystems/blueprint/ModalComponent.tsx:80]: Moved this to document.body

**Review by somangshu — DISMISSED** [review]
LGTM, Tested on Deploy preview as well. 

@marks0351 Kindly review.

File: app/client/src/index.css [hunk 4]
```diff
@@ -93,3 +93,12 @@ div.bp3-popover-arrow {
   background: white !important;
   border-bottom: 1px solid #E7E7E7 !important;
 }
+/* making both the Modal layer and the Dropdown layer */
+.bp3-modal-widget, .appsmith_widget_0 > .bp3-portal {
+  z-index: 2 !important;
```
Comment (marks0351) [inline app/client/src/index.css:98]: @Tooluloope can u pls add these z-index's to the layers as well.

**Review by marks0351 — CHANGES_REQUESTED** [review]

File: app/client/src/index.css (hunk 4 above)
Comment (Tooluloope) [reply app/client/src/index.css:98]: Done

**Review by marks0351 — APPROVED** [review]

**Comment by somangshu** [comment]
This PR is blocked because of unrelated failing test

**Review by riodeuno — APPROVED** [review]

**Comment by somangshu** [comment]
@Tooluloope related test failing here!

**Comment by somangshu** [comment]
@marks0351 need you approval here

**Review by marks0351 — APPROVED** [review]

**Comment by somangshu** [comment]
@Tooluloope please resolve the merge conflicts 👆

**Review by somangshu — APPROVED** [review]

**Comment by somangshu** [comment]
@Tooluloope merge required to fix failing tests

**Review by somangshu — APPROVED** [review]
Approving to run tests

**Pull request merged**